README.md
# anton

A quick and dirty air sensor which submits data directly to an [InfluxDB][]
v1 instance, without auth.

I built this because the west coast is on fire, and I ran out of [Luftdaten][]
sensor parts. It's not pretty or full featured but it works.

[InfluxDB]: https://docs.influxdata.com/influxdb/v1.8/
[Luftdaten]: https://sensor.community/

## Requirements

Hardware:

* An ESP development board for ESP8266 or ESP32, such as:
  * NodeMCUv3 or similar ESP8266 platform
  * Lolin32 or similar ESP32 platform
* A particulate sensor, one of:
  * [ZH03B][]
  * [PMS7003][]; Only the PMS7003 has been tested, but other PMSX003 sensors may
    work.
* (Optional) A multisensor:
  * [BME680][] sensor and breakout board; configured for i2c
* (Optional) A CO₂ sensor:
  * [MH-Z19B][]

[ZH03B]: https://www.winsen-sensor.com/sensors/dust-sensor/zh3b.html
[PMS7003]: http://www.plantower.com/en/content/?110.html
[BME680]: https://www.bosch-sensortec.com/products/environmental-sensors/gas-sensors-bme680/
[MH-Z19B]: https://www.winsen-sensor.com/sensors/co2-sensor/mh-z19b.html

Software:

* InfluxDB v1.8

Wire the ZH03B as follows:

* Pin1: `VU`
* Pin2: `Ground`
* ESP8266
  * Pin4: `D4`
  * Pin5: `D3`
* ESP32
  * Pin4: `TX2` (`GPIO17`)
  * Pin5: `RX2` (`GPIO16`)

Or the PMS7003 as follows:

* Pin1: `VU`
* Pin3: `Ground`
* ESP8266
  * Pin7: `D3`
  * Pin9: `D4`
* ESP32
  * Pin7: `TX2` (`GPIO17`)
  * Pin9: `RX2` (`GPIO16`)

(Optional) Wire the BME680 as follows:

* VCC: `3v3`
* GND: `Ground`
* ESP8266
  * SCL: `D1`
  * SDA: `D2`
* ESP32
  * SCL: `SCL` (`GPIO22`)
  * SDA: `SDA` (`GPIO21`)

(Optional) Wire the MH-Z19B as follows, tested only on ESP32:

* VCC: `VU`
* GND: `Ground`
* ESP8266
  * TXD: `D5`
  * RXD: `D6`
* ESP32:
  * TXD: `GPIO18`
  * RXD: `GPIO19`

A 3D printable case for Anton is available at [PrusaPrinters][]. **Note:** the
case fits the ZH03B and ESP8266; for PMS7003 or ESP32, you're on your own right
now, but I'll make one in the future.

[PrusaPrinters]: https://www.prusaprinters.org/prints/40746-case-for-anton-air-quality-influxdb

## Installation

This is built using [PlatformIO][]; refer to their docs on building/uploading.

[PlatformIO]: https://platformio.org/

## Setup

Once uploaded and booted, a wireless network called `anton` will be
created; join this network and you will be directed to a captive portal for
configuration.

## Measurements

Measurements are submitted to the configured database as a measurement named
`particulate_matter`. It will contain the following tags:

* `node` the sensor name
* `location` (optional) the sensor location

…and the following fields containing integer values:

* `p1_0` PM1.0
* `p2_5` PM2.5
* `p10_0` PM10
* `aqi` EPA [AQI][]

[AQI]: https://www.airnow.gov/aqi/aqi-basics/

…and the following string values:

* `aqi_contributor` the primary AQI contributor (highest value); the string will
  match the measurement name, e.g. `p2_5`

**Note:** currently only PM10 and PM2.5 are accounted for in AQI measurements,
following the EPA standard.

**Note:** if the AQI cannot be determined, either due to overflow or other
error, it will be omitted from the InfluxDB submission.

### Multisensor

If you have the optional multisensor connected, the following additional fields
fields will be sent; all are integers unless noted:

* `temperature_c` (float) °C
* `temperature_c_raw` (float) °C, uncompensated
* `humidity` %rh
* `pressure` hPa
* `gas_resistance` Ohm
* `iaq` EPA [IAQ][].
* `iaq_accuracy` IAQ accuracy; see notes
* `co2_equiv` Equivalent CO₂ PPM
* `co2_equiv_accuracy` Equivalent CO₂ accuracy; see notes
* `breath_voc` (float) Breath VOC
* `breath_voc_accuracy` Breath VOC Accuracy; see notes

[IAQ]: https://www.epa.gov/indoor-air-quality-iaq/introduction-indoor-air-quality

**Note:** Various `accuracy` values are provided by the Bosch Sensortec
[BSEC][bsec] library, which have the following meaning ([source][bsec-src]):

* `0` sensor is stabilizing
* `1` readings were too stable to define references
* `2` currently calibrating
* `3` calibrated

[bsec-src]: https://community.bosch-sensortec.com/t5/Question-and-answers/What-does-the-IAQ-accuracy-mean-in-BSEC/qaq-p/5935

### Dedicated CO₂ Sensor

If you have the optional dedicated CO₂ sensor connected, the following
additional integer fields will be sent:

* `co2` PPM

**Note:** when using the optional dedicated CO₂ sensor, you must provide a base
calibration every few weeks. This can be done from the dashboard's calibration
page, and should be initiated when background levels are low. If indoors it's
recommended to open a window and allow good air circulation for several minute
before performing the calibration. This calibration takes 20 minutes.

## Administration

Once configured, a webserver is exposed on port 80 which can be used to view
sensor values, update firmware, change configuration, and reboot or reset the
sensor.

## Changelog

See [CHANGELOG](./CHANGELOG.md).

## Acknowledgements

This was quick to implement due to the following excellent libraries:

* [@ShaggyDog18/SD_ZH03B](https://github.com/ShaggyDog18/SD_ZH03B)
* [@tobiasschuerg/InfluxDB-Client-for-Arduino](https://github.com/tobiasschuerg/InfluxDB-Client-for-Arduino)
* [@prampec/IotWebConf](https://github.com/prampec/IotWebConf)
* [@BoschSensortec/BSEC-Arduino-library][bsec]
* [@fu-hsi/PMS](https://github.com/fu-hsi/PMS)
* [@Erriez/ErriezMHZ19B](https://github.com/Erriez/ErriezMHZ19B)

## License

The code contained in this repo is released under the GNU [GPL-3.0](./LICENSE).

Building this project requires pulling in the various submodules under the `lib`
directory, each of which has their own license; of note the [BSEC][bsec] library
provided by [Bosch Sensortec][bosch] has a closed source binary blob provided
under a unique license which you must abide to use this software.

[bsec]: https://github.com/BoschSensortec/BSEC-Arduino-library
[bosch]: https://www.bosch-sensortec.com/software-tools/software/bsec/


CHANGELOG.md
# Changelog

Changes which are not backwards compatible will be listed here. Anton is under
active development and does not have a fixed feature set, nor does it have
version numbers yet. For now, changes will be listed by date.

## 2022-01-03

To report more accurate temperatures, the following changes were made:

* Temperature fields are now floats
* A compile-time temperature offset may be provided, which is used by the BSEC
  library to perform its calculations.

### Temperature Fields

Temperature fields are now floats; in previous versions they were integers,
which loses a lot of precision when dealing in Celsius. To avoid conflicts
with existing databases, fields were renamed/added:

* `temperature` is now `temperature_c`
* `temperature_c_raw` is the uncompensated temperature; direct from the sensor

**Note:** `temperature_c_raw` is _not_ the same as "`temperature_c` ignoring
offset", but rather the raw reading from the sensor itself. If you needed the
previous equivalent `temperature` you should take `temperature_c` and remove the
compiled offset from it, although there shouldn't be much use for this value.

### Compile-time Temperature Offset

When compiling Anton firmware, you may define the float value
`BSEC_TEMPERATURE_OFFSET_C` which will be passed to the underlying [BSEC][bsec]
library to determine the temperature offset that should be used. The default
value of `4.0` works well for various 3D printable cases, and you may override
it for your application as necessary.

**Note:** This value is _subtracted_ from the values that are read; you will
almost always wish to specify a positive value here.

## 2021-12-12

Move to Bosch Sensortec's [BSEC][bsec] library for the BME680, which improves
accuracy of the temperature/humidity readings as well as provides equivalent CO₂
values without the dedicated sensor. However this brings additional license
restrictions which you should [view and accept][bosch] before using.

[bsec]: https://github.com/BoschSensortec/BSEC-Arduino-library
[bosch]: https://www.bosch-sensortec.com/software-tools/software/bsec/


## 2021-09-25

Does away with pin/UART configurations and makes them compile time. Also added
CO₂ sensor options. This caused a revision of the configuration options, and you
will need to reconfigure the sensor after flashing.

## 2021-07-04

Moves from [WifiManager][] to [IotWebConf][]. *This is a backwards incompatible
change.* You will need to reconfigure the sensor after upgrading.

[WifiManager]: https://github.com/tzapu/WiFiManager
[IotWebConf]: https://github.com/prampec/IotWebConf

## 2020-10-22

Moved from the deprecated `SPIFFS` to the still-maintained `LittleFS`. *This is
a backwards incompatible change.* You will need to reconfigure the sensor after
upgrading.

## 2020-10-03

Earlier versions of Anton submitted values to InfluxDB as floats; this has been
corrected to use integer values. Updating will require you to either use a new
database or rewrite your existing one.


src/main.cpp
#include <Arduino.h>

#include <AQI.h>

#include <IotWebConf.h>
#include <IotWebConfTParameter.h>

#include "strings.h"

#include "NetworkController.h"
#include "SensorController.h"
#include "WebServerController.h"
#include "reporters/Reporter.h"
#include "reporters/InfluxDB_Reporter.h"
#include "sensors/AirSensor.h"
#include "sensors/ZH03B_AirSensor.h"
#include "sensors/PMS_AirSensor.h"
#include "sensors/EnvironmentSensor.h"
#include "sensors/BME680_EnvironmentSensor.h"
#include "sensors/CO2Sensor.h"
#include "sensors/MHZ19B_CO2Sensor.h"
#include "util.h"

#ifndef GIT_REV
#define GIT_REV "Unknown"
#endif

#ifdef ESP8266
#define ANTON_PLATFORM "ESP8266"
#elif defined(ESP32)
#define ANTON_PLATFORM "ESP32"
#endif

#define PRODUCT_NAME "anton"
#define DEFAULT_PASSWORD "anton-system"

#ifndef BSEC_TEMPERATURE_OFFSET_C
#define BSEC_TEMPERATURE_OFFSET_C 4.0
#endif

// pins
#ifdef ESP8266
#define PARTICULATE_RX D3
#define PARTICULATE_TX D4
// TODO: reevaluate these
#define CO2_RX D5
#define CO2_TX D6
#elif defined(ESP32)
#define PARTICULATE_UART 2
#define CO2_UART 1
#define UART_1_RX 18
#define UART_1_TX 19
#endif

DNSServer dnsServer;
WebServer webServer(80);

Stream *airSensorSerial = nullptr;
AirSensor *airSensor = nullptr;
Reporter *reporter = nullptr;
EnvironmentSensor *environmentSensor = nullptr;
Stream *co2SensorSerial = nullptr;
CO2Sensor *co2Sensor = nullptr;

SensorController *sensorController = nullptr;
NetworkController *networkController = nullptr;
WebServerController *webserverController = nullptr;

AntonConfiguration config;

void setup()
{
  Serial.begin(115200);
  Serial.printf("setup: starting version %s\n", GIT_REV);

  networkController = new NetworkController(PRODUCT_NAME, DEFAULT_PASSWORD, &dnsServer, &webServer);

  config = networkController->getConfig();

  // set up reporter
  InfluxDB_Reporter *DB = new InfluxDB_Reporter(config.sensorName,
                                                config.sensorLocation,
                                                config.influxdbUrl,
                                                config.influxdbDatabase);
  reporter = DB;

  if (config.particleSensorEnabled)
  {
#ifdef ESP8266
    airSensorSerial = util::getSerial(PARTICULATE_RX, PARTICULATE_TX);
#elif defined(ESP32)
    airSensorSerial = util::getSerial(PARTICULATE_UART);
#endif

    if (config.particleSensor == ParticleSensor::Winsen_ZH03B)
    {
      Serial.println("setup: starting ZH03B particle sensor on serial");
      ZH03B_AirSensor *ZH = new ZH03B_AirSensor(*airSensorSerial);
      airSensor = ZH;
    }
    else if (config.particleSensor == ParticleSensor::Plantower_PMS7003)
    {
      Serial.println("setup: starting PMS7003 particle sensor on serial");
      PMS_AirSensor *PMS = new PMS_AirSensor(*airSensorSerial);
      airSensor = PMS;
    }
    else
    {
      Serial.printf("setup: invalid particle sensor type: %d\n", config.particleSensor);
    }
  }

  if (config.vocSensorEnabled)
  {
    Serial.println("setup: starting BME680 VOC sensor on I2C (default pins)");
    Wire.begin();
    BME680_EnvironmentSensor *BME = new BME680_EnvironmentSensor(Wire, BSEC_TEMPERATURE_OFFSET_C);
    environmentSensor = BME;
  }

  if (config.co2SensorEnabled)
  {
#ifdef ESP8266
    co2SensorSerial = util::getSerial(CO2_RX, CO2_TX);
#elif defined(ESP32)
    co2SensorSerial = util::getSerial(CO2_UART, 9600, SERIAL_8N1, UART_1_RX, UART_1_TX);
#endif
    co2Sensor = new MHZ19B_CO2Sensor(co2SensorSerial);
  }

  sensorController = new SensorController(reporter, airSensor, co2Sensor, environmentSensor);
  webserverController = new WebServerController(&webServer, networkController, sensorController);
}

void loop()
{
  networkController->loop();
  webserverController->loop();
  sensorController->loop();
}
